server/service/service.go

```diff
@@ -218,6 +219,104 @@ const (
 	stateAttested
 )
 
+// buildLookupFromReportStatus constructs an EntityLookup object from a ReportStatusRequest.
+func buildLookupFromReportStatus(ctx context.Context, req *bpb.ReportStatusRequest) (*types.EntityLookup, string, error) {
+	peerAddr, err := peerAddressFromContext(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+	var ccSerial string
+	if states := req.GetStates(); len(states) > 0 {
+		ccSerial = states[0].GetSerialNumber()
+	}
+	if ccSerial == "" {
+		return nil, "", status.Errorf(codes.InvalidArgument, "control card serial is missing in ReportStatusRequest")
+	}
+	// Note: ReportStatusRequest doesn't have manufacturer or part number. The lookup is simpler.
+	lu := &types.EntityLookup{
+		SerialNumber: ccSerial,
+		IPAddress:    peerAddr,
+	}
+	return lu, ccSerial, nil
+}
+
+// sendIdevidChallenge contains the logic for parsing an IDevID cert, and sending a nonce challenge.
+func (s *Service) sendIdevidChallenge(stream bpb.Bootstrap_BootstrapStreamServer, session *streamSession, deviceID string, idevidCertB64 string) error {
+	certDER, err := base64.StdEncoding.DecodeString(idevidCertB64)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "failed to decode idevid_cert: %v", err)
+	}
+	cert, err := x509.ParseCertificate(certDER)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "failed to parse idevid_cert: %v", err)
+	}
+
+	// TODO: Validate this cert against Vendor CAs & check against Entity Manager.
+	// e.g., s.em.ValidateIDevID(ctx, cert, chassis)
+	session.idevidCert = cert
+	log.Infof("Successfully parsed IDevID cert for %s", cert.Subject.CommonName)
+
+	nonceBytes := make([]byte, 32)
+	if _, err := rand.Read(nonceBytes); err != nil {
+		return status.Errorf(codes.Internal, "failed to generate nonce: %v", err)
+	}
+	session.nonce = nonceBytes
+	nonceStr := base64.StdEncoding.EncodeToString(nonceBytes)
+
+	challengeResp := &bpb.BootstrapStreamResponse{
+		Type: &bpb.BootstrapStreamResponse_Challenge_{
+			Challenge: &bpb.BootstrapStreamResponse_Challenge{
+				Type: &bpb.BootstrapStreamResponse_Challenge_Nonce{
+					Nonce: nonceStr,
+				},
+			},
+		},
+	}
+	if err := stream.Send(challengeResp); err != nil {
+		return err
+	}
+	session.currentState = stateTPM20NonceSent
+	log.Infof("Sent nonce challenge to IDevID device: %s", deviceID)
+	return nil
+}
+
+// establishSessionAndSendChallenge is a helper to establish session and send challenge for TPM2.0 with IdevID.
+func (s *Service) establishSessionAndSendChallenge(stream bpb.Bootstrap_BootstrapStreamServer, ctx context.Context, lookup *types.EntityLookup, identity *bpb.Identity, ccSerial string) (*streamSession, string, error) {
```

Comment:
Instead of having this function return a new streamSession, could you update it so that it simply updates the existing one? It might save us from some complicated bugs if we just make sure that we create one session in the first connection and then update it as the stream happens.

Reply:
ok, that makes sense.